### PR TITLE
feat: support BackendRef HTTP filters

### PR DESCRIPTION
**Pull request by cnvergence**
**What type of PR is this?**

**What this PR does / why we need it**:
- Support HTTP Filters in the BackendRef and translate them to WeightedCluster in the XDS.
   Currently, only support for RequestHeaderModifiier and ResponseHeaderModifier type, as it is the only filter with 
   straightforward support in
   https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-weightedcluster-clusterweight
- Validate other types of filters

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/2572 https://github.com/envoyproxy/gateway/issues/3338

### CODE REVIEW

**Comment by cnvergence** [comment]
@arkodg initial PTAL, I have used weighted cluster, but for the use case mentioned in the issue, we would need to create a separate weighted cluster per backend

**Comment by zirain** [comment]
this may conflict with https://github.com/envoyproxy/gateway/pull/3212, should decide which go first.
cc @envoyproxy/gateway-maintainers

**Comment by cnvergence** [comment]
@zirain rebased the PR, as your changes went first 😄

File: internal/gatewayapi/route.go
Comment (shawnh2) [inline internal/gatewayapi/route.go:1236]: can we add a nil check here just to avoid all the `filters: {}` changes in this PR?

File: internal/gatewayapi/route.go
Comment (arkodg) [inline internal/gatewayapi/route.go:1249]: return `nil` here ?

File: internal/gatewayapi/route.go
Comment (arkodg) [inline internal/gatewayapi/route.go:1254]: or len == 0?

File: internal/gatewayapi/route.go
Comment (arkodg) [reply internal/gatewayapi/route.go:1236]: +1

File: internal/gatewayapi/route.go
Comment (arkodg) [inline internal/gatewayapi/route.go:1264]: why this is additional variable needed ?

File: internal/xds/translator/route.go
Comment (arkodg) [inline internal/xds/translator/route.go:227]: nice !

File: internal/xds/translator/route.go
Comment (arkodg) [inline internal/xds/translator/route.go:624]: prefer if this could be simplified to `filter != nil` only

File: test/conformance/conformance_test.go
Comment (arkodg) [inline test/conformance/conformance_test.go:27]: not tied to this PR, but can we fix the link, its https://github.com/envoyproxy/gateway/issues/3352

**Comment by arkodg** [comment]
hey @cnvergence small nits, overall LGTM, thanks for adding this feature !
you will also want to undo parts of https://github.com/envoyproxy/gateway/pull/2620 in this PR

File: internal/gatewayapi/route.go
Comment (cnvergence) [reply internal/gatewayapi/route.go:1264]: not at all, thanks

File: internal/gatewayapi/route.go
Comment (cnvergence) [reply internal/gatewayapi/route.go:1236]: sorry, problem was withing getBackendFilters func

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: internal/xds/translator/route.go
Comment (zhaohuabing) [inline internal/xds/translator/route.go:226]: Nit: this comment should be updated. Adding a few lines to explain the reason would help people who come later.

**Review by zhaohuabing — COMMENTED** [review]
This PR looks great! Just a few nits.

**Review by zhaohuabing — APPROVED** [review]

**Pull request merged**
